src/TreeSitter/Node.hs

```diff
@@ -20,70 +26,107 @@ data Node = Node
   , nodeNamedChildCount :: !Int32
   , nodeChildCount :: !Int32
   }
-  deriving (Show, Eq, Generic, CStorable)
+  deriving (Show, Eq, Generic)
 
 data TSPoint = TSPoint { pointRow :: !Int32, pointColumn :: !Int32 }
-  deriving (Show, Eq, Generic, CStorable)
+  deriving (Show, Eq, Generic)
 
+data TSNode = TSNode !(Ptr ()) !Word32 !Word32
+  deriving (Show, Eq, Generic)
 
-data TSNode = TSNode !(Ptr ()) !Int32 !Int32 !Int32
-  deriving (Show, Eq, Generic, CStorable)
 
-peekAdvance :: forall a b. Storable a => Ptr a -> IO (a, Ptr b)
-peekAdvance ptr = do
-  let aligned = alignPtr ptr (alignment (undefined :: a))
+-- | 'Struct' is a strict 'Monad' with automatic alignment & advancing, & inferred type.
+newtype Struct a = Struct { runStruct :: forall b . Ptr b -> IO (a, Ptr a) }
+
+evalStruct :: Struct a -> Ptr b -> IO a
+evalStruct = fmap (fmap fst) . runStruct
+
+peekStruct :: forall a . Storable a => Struct a
+peekStruct = Struct (\ p -> do
+  let aligned = alignPtr (castPtr p) (alignment (undefined :: a))
   a <- peek aligned
-  return (a, castPtr aligned `plusPtr` sizeOf a)
-{-# INLINE peekAdvance #-}
+  pure (a, aligned `plusPtr` sizeOf a))
+{-# INLINE peekStruct #-}
 
-pokeAdvance :: forall a b. Storable a => Ptr a -> a -> IO (Ptr b)
-pokeAdvance ptr a = do
-  let aligned = alignPtr ptr (alignment (undefined :: a))
+pokeStruct :: Storable a => a -> Struct ()
+pokeStruct a = Struct (\ p -> do
+  let aligned = alignPtr (castPtr p) (alignment a)
   poke aligned a
-  return (castPtr aligned `plusPtr` sizeOf a)
-{-# INLINE pokeAdvance #-}
+  pure ((), castPtr aligned `plusPtr` sizeOf a))
+{-# INLINE pokeStruct #-}
```

Comment:
Inlining these means that `Struct` may be slightly faster than the `peekAdvance`/`pokeAdvance` functions which predated them.